### PR TITLE
MM-10272: Add a count for pinned posts header icon

### LINKED ISSUE #10145: Add a count for pinned posts header icon
If you're interested please comment here and come [join our "Contributors" community channel](https://community.mattermost.com/core/channels/tickets) on our daily build server, where you can discuss questions with community members and the Mattermost core team. For technical advice or questions, please  [join our "Developers" community channel](https://community.mattermost.com/core/channels/developers).

New contributors please see our [Developer's Guide](https://developers.mattermost.com/contribute/getting-started/).

----

**Notes**: [Jira ticket](https://mattermost.atlassian.net/browse/MM-10272)

This improvement involves both client-side (ReactJS), server-side (Go) and mobile (React Native) changes:

*Client-side*

Add a count, similar to the channel member icon, to the pinned post icon in the channel header.

The count should increment when a post is pinned and decrement when a post is unpinned.

image

*Server-side*

There are a few implementation options to consider in order to ensure this change won't decrease server performance. We advice loadtesting server-side changes using our [loadtest tool|https://github.com/mattermost/mattermost-load-test].

1. Use a SELECT query the Posts table to update the pinned posts count.
2. Store the pin count in the Channel Props? When a post is pinned then the count increased, when deleted decreased. That way no dedicated SQL query is required. Requires a database migration.
3. Create a separate table for tracking pinned posts. Requires a database migration.

Note that we already send a WebSocket event when a post gets pinned, since users viewing a post getting pinned in a channel see the update in real time.

*React Native*

image

### PR DESCRIPTION
**Pull request by jkl5616**
#### Summary
As part of #10145, added `pinnedpost_count` to `/api/v4/channels/:channel_id/stats` for counting pinned posts.

#### Ticket Link
#10145

### CODE REVIEW

File: api4/channel.go [hunk 1]
```diff
@@ -527,7 +527,13 @@ func getChannelStats(c *Context, w http.ResponseWriter, r *http.Request) {
 		return
 	}
 
-	stats := model.ChannelStats{ChannelId: c.Params.ChannelId, MemberCount: memberCount, GuestCount: guestCount}
+	pinnedPostCount, err := c.App.GetChannelPinnedPostCount(c.Params.ChannelId)
```
Comment (saturninoabril) [inline api4/channel.go]: Please add test to `TestGetChannelStats` of `api4/channel_test.go` that validates `stats.PinnedPostCount`?

Comment (jkl5616) [reply api4/channel.go]: @saturninoabril Please check for the changes!

File: api4/channel_test.go [hunk 2]
```diff
@@ -1849,6 +1849,8 @@ func TestGetChannelStats(t *testing.T) {
 		t.Fatal("couldnt't get extra info")
 	} else if stats.MemberCount != 1 {
 		t.Fatal("got incorrect member count")
+	} else if stats.PinnedPostCount != 0 {
+		t.Fatal("got incorrect pinned post count")
```
Comment (saturninoabril) [inline api4/channel_test.go]: Thanks for adding but could you add another case where there is at least a number of pinned post/s?

File: api4/channel.go (hunk 1 above)
Comment (saturninoabril) [reply api4/channel.go]: One more request and I'm all good - https://github.com/mattermost/mattermost-server/pull/11840#discussion_r315082010

File: api4/channel_test.go (hunk 2 above)
Comment (jkl5616) [reply api4/channel_test.go]: @saturninoabril I added the test!

Comment (saturninoabril) [reply api4/channel_test.go]: Thanks!

**Comment by jasonblais** [comment]
@mkraft Friendly reminder for helping review this PR :)

**Comment by mkraft** [comment]
Removing myself as there have been two dev reviews.

**Comment by justinegeffen** [comment]
The Jira ticket for this indicates that the milestone is now 19 December 2019: https://mattermost.atlassian.net/browse/MM-10272. I have taken a look on https://mysql.test.mattermost.com/open-testing/channels/town-square (release-5.16-1141) and can't locate the count, nor the pin next to the flag. 
Screen Shot 2019-10-15 at 10 52 36

Should I be looking at a different server, or has this been moved out as per the Jira ticket?

**Comment by jasonblais** [comment]
@justinegeffen that's correct, looks like the Jira ticket has been moved to 5.18

@jkl5616 I noticed you submitted a PR for the mobile project (https://github.com/mattermost/mattermost-mobile/pull/3290), are any changes needed for webapp for the pinned posts count appear on the UI?

**Comment by esethna** [comment]
Yes I believe the webapp PR is still needed. @jkl5616 let us know if you have any questions on the implementation?

**Comment by esethna** [comment]
@jkl5616 would you be interested in working on adding the pinned count to the webapp? Let us know if we can provide some guidance?

image

**Comment by justinegeffen** [comment]
@amyblais, will docs not be needed at all for this or are they on hold pending another PR?

**Comment by amyblais** [comment]
@justinegeffen It's missing a webapp PR before this is complete.

**Comment by justinegeffen** [comment]
Got it, thanks.

**Pull request merged**